Feature/4 projects cleanup

kwraith05 opened the pull request:
This PR cleans up the 4-projects folder and aligns it with the structure based on environments and business units.

kwraith05 commented:
@bharathkkb @morgante @rjerrems Please take a look and let me know of any changes I should make.

daniel-cit commented on 4-projects/modules/standard_projects/data.tf:
```diff
@@ -31,3 +31,12 @@ data "google_compute_network" "nonprod_shared_vpc" {
 }
+
+data "google_projects" "dev_project" {
+  filter = "labels.application_name:org-shared-vpc-dev"
```
The values for `labels.application_name` and `google_compute_network.name` used in this file are no longer valid.

Current values can be found in the files below:

-  https://github.com/terraform-google-modules/terraform-example-foundation/blob/develop/2-environments/modules/env_baseline/networking.tf
- https://github.com/terraform-google-modules/terraform-example-foundation/blob/develop/3-networks/modules/restricted_shared_vpc/main.tf
- https://github.com/terraform-google-modules/terraform-example-foundation/blob/develop/3-networks/modules/standard_shared_vpc/main.tf

rjerrems replied:
This feedback is still valid for the `single_project` module

bharathkkb commented on 4-projects/business_unit_2/dev/example_single_project.tf:
```diff
@@ -0,0 +1,31 @@
+  folder_id = var.parent_folder
+
+  # Metadata
+  project_prefix   = "sample-single-bu2-d"
```
Please correct me if I am wrong but looking at the white-paper, shouldn't project names be prefixed with `prf-`?

daniel-cit replied:
Nice catch.
Actually it is `prj-`.  
I think that in this case it should be `prj-bu2-d-sample-single`

rjerrems replied:
Yep, we should rename project with that naming convention

bharathkkb commented on 4-projects/business_unit_2/dev/example_single_project.tf:
```diff
@@ -0,0 +1,31 @@
+  org_id                      = var.org_id
+  billing_account             = var.billing_account
+  impersonate_service_account = var.terraform_service_account
+  environment                 = "dev"
```
if we are passing in `environment` here the module should be able to ascertain the env short code `d` without explicitly passing in `-d` for `project_prefix`

bharathkkb commented on 4-projects/modules/standard_projects/main.tf:
```diff
@@ -64,6 +67,27 @@ module "prod_project" {
 
+module "dev_project" {
+  source                      = "terraform-google-modules/project-factory/google"
+  version                     = "~> 7.0"
```
bump to 8.0

rjerrems replied:
+1 - if we can do this for all projects that would be great

bharathkkb commented on 4-projects/modules/standard_projects/main.tf:
```diff
@@ -64,6 +67,27 @@ module "prod_project" {
+  org_id                      = var.org_id
+  billing_account             = var.billing_account
+  folder_id                   = var.dev_folder_id
+
```
expose `skip_gcloud_download ` here and though out with default to `true`

rjerrems replied:
+1 - if we can do this for all projects that would be great

bharathkkb commented on 4-projects/business_unit_2/dev/example_single_project_optional.tf:
```diff
@@ -0,0 +1,45 @@
+#   application_name = "sample-single-project-app-optional-bu2-d"
+
+#   # Network Setting (Optional)
+#   enable_networking    = true
```
I think we should have some of these optionals ( and optional modules) enabled in some env so that we can test these code paths as well. @rjerrems WDYT?

rjerrems replied:
Yeah this is a good idea, although the DNS module will need some updates with the new DNS hub architecture, but we can deal with that separately.

rjerrems commented on 4-projects/business_unit_1/dev/backend.tf:
```diff
@@ -0,0 +1,22 @@
+/**
```
+1

rjerrems commented on 4-projects/modules/folder_environments/README.md:
```diff
@@ -10,6 +10,7 @@
 
 | Name | Description |
 |------|-------------|
+| dev\_folder\_id |  |
```
The folder environments module should probably be removed now, as in this update to the code base the environment folder is created at the top of the hierarchy in the 2-environment steps

rjerrems commented on 4-projects/modules/standard_projects/data.tf:
```diff
@@ -14,20 +14,56 @@
 
-data "google_projects" "prod_project" {
-  filter = "labels.application_name:org-shared-vpc-prod"
+data "google_projects" "base_prod_project" {
```
Similar to the folder_environments, the standard projects module makes less sense in the new structure where we have sub folders per environment. I think we can remove it

kwraith05 commented:
I've addressed the above comments and:
- deleted folder environments and standard projects modules
- updated the example projects to use the appropriate naming conventions
- bumped the projects versions to 8.0
- exposed the skip_gcloud_download variable and set the default to true for all projects

@bharathkkb @daniel-cit @rjerrems  - Please review and let me know of any further edits you'd like me to make

rjerrems commented on 4-projects/business_unit_1/dev/example_single_project.tf:
```diff
@@ -0,0 +1,38 @@
+
+locals {
+  env           = "dev"
+  env_code      = element(split("", local.env), 0)
```
it might make sense to do this logic in the single project module, to keep things dry

rjerrems commented on 3-networks/envs/dev/main.tf:
```diff
@@ -27,15 +27,15 @@ locals {
 
 data "google_projects" "restricted_host_project" {
-  filter = "labels.application_name=restricted-shared-vpc-host-${local.env}  lifecycleState=ACTIVE"
+  filter = "labels.application_name=restricted-shared-vpc-host-${local.env}"
```
I think these have been reverted from the changes that @daniel-cit made, can we change it back?

rjerrems commented on 4-projects/business_unit_1/dev/example_single_project.tf:
```diff
@@ -0,0 +1,42 @@
+
+  # Metadata
+  project_prefix    = "prj-${local.business_code}-${var.env_code}-sample-single"
+  cost_centre       = "cost-centre-1"
```
Can we remove `cost_centre` as its a duplicate of the billing code field

rjerrems commented on 4-projects/business_unit_1/dev/example_single_project_optional.tf:
```diff
@@ -0,0 +1,57 @@
+
+# locals {
+#   env           = "dev"
+#   env_code      = element(split("", local.env), 0)
```
we should remove these locals, if someone un-commented this it would fail as they are duplicates of what is defined in the other file.

rjerrems commented on 4-projects/business_unit_1/nonprod/example_single_project.tf:
```diff
@@ -0,0 +1,42 @@
+  folder_id = var.parent_folder
+
+  # Metadata
+  project_prefix    = "prj-${local.business_code}-${var.env_code}-sample-single"
```
this can be generated in the single_project module, given you are supplying `business_code` and `env_code` as variables already. (which will help standardize the naming convention)

rjerrems commented on 4-projects/modules/single_project/main.tf:
```diff
@@ -20,22 +20,28 @@ locals {
+    primary_contact   = var.primary_contact
+    secondary_contact = var.secondary_contact
+    business_code     = var.business_code
+    env_code          = element(split("", var.environment), 0)
```
This isn't in the white paper, so we can probably remove this as a label, given environment is already there.

rjerrems commented on 4-projects/modules/single_project/main.tf:
```diff
@@ -20,22 +20,28 @@ locals {
   random_project_id           = "true"
   impersonate_service_account = var.impersonate_service_account
   activate_apis               = var.activate_apis
   name                        = "${var.project_prefix}-${var.environment}"
```
we can remove environment from the name here as we are defining as env_code is already in the project name

rjerrems commented on 4-projects/business_unit_1/dev/example_single_project.tf:
```diff
@@ -0,0 +1,42 @@
+  # Metadata
+  project_prefix    = "prj-${local.business_code}-${var.env_code}-sample-single"
+  cost_centre       = "cost-centre-1"
+  application_name  = "sample-single-prj-app-${local.business_code}-${var.env_code}"
```
Application name is supposed to be used as something that allows you to identify all projects for a given application. For example, you could filter all projects that have the application name  "splunk" and it would return projects across all environments. Eg:

`gcloud projects list --filter="labels.application_name=splunk"`

This also lets you see the cost of a single application across environments. So this should probably become something simpler, that is not unique across environments. For example:

<business_code>-sample-single

rjerrems commented on 4-projects/business_unit_1/prod/variables.tf:
```diff
@@ -0,0 +1,65 @@
+}
+
+variable "vpc_type" {
+  description = "The type of VPC to attach the project to (base, rest, etc)"
```
nit: this variable should tell you all the possible options and preferably is not abbreviated, to improve readability
